### PR TITLE
CSHARP-1541: Add support for $comment for profiling

### PR DESCRIPTION
**Pull request by BorisDog**
[EG](https://spruce.mongodb.com/version/6213f18f5623435871e5c99)

Done together with [CSHARP-4043](https://jira.mongodb.org/browse/CSHARP-4043)
Follow up work in [CSHARP-4067](https://jira.mongodb.org/browse/CSHARP-4067)

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/UnifiedTestOperations/Matchers/UnifiedValueMatcher.cs [hunk 1]
```diff
@@ -158,7 +158,7 @@ private void AssertValuesMatch(BsonValue actual, BsonValue expected, bool isRoot
 
                 for (int i = 0; i < expectedArray.Count; i++)
                 {
-                    AssertValuesMatch(actualArray[i], expectedArray[i], isRoot: false);
+                    AssertValuesMatch(actualArray[i], expectedArray[i], isRoot: isRoot && !isRecursiveCall);
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/UnifiedTestOperations/Matchers/UnifiedValueMatcher.cs:161]: We need to allow "actual" documents in root array to have additional fields. [Spec](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#allowing-extra-fields-in-root-level-documents).

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs [hunk 2]
```diff
@@ -34,6 +34,7 @@ public class CreateIndexesOperation : IWriteOperation<BsonDocument>
         // fields
         private readonly CollectionNamespace _collectionNamespace;
         private CreateIndexCommitQuorum _commitQuorum;
+        private BsonValue _comment;
```
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs:37]: Also, what about EndTransactionOperation (Abort/Commit). It's not main operation, but the same story is in `AggregateToCollection` where we added `$comment`

[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs:37]: Also what about getMore in cursors?

File: src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs [hunk 3]
```diff
@@ -194,7 +204,8 @@ private ReadCommandOperation<BsonDocument> CreateOperation()
                 { "filter", _filter, _filter != null },
                 { "nameOnly", () => _nameOnly.Value, _nameOnly.HasValue },
                 { "cursor", () => new BsonDocument("batchSize", _batchSize.Value), _batchSize.HasValue },
-                { "authorizedCollections", () => _authorizedCollections.Value, _authorizedCollections.HasValue }
+                { "authorizedCollections", () => _authorizedCollections.Value, _authorizedCollections.HasValue },
+                { "comment", () => _comment, _comment != null },
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs:208]: let's not add delegate if it's not needed

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs:208]: `,` in the end

File: src/MongoDB.Driver.Core/Core/Operations/ListDatabasesOperation.cs [hunk 4]
```diff
@@ -133,7 +146,8 @@ internal BsonDocument CreateCommand()
                 { "listDatabases", 1 },
                 { "filter", _filter, _filter != null },
                 { "nameOnly", _nameOnly, _nameOnly != null },
-                { "authorizedDatabases", _authorizedDatabases, _authorizedDatabases != null }
+                { "authorizedDatabases", _authorizedDatabases, _authorizedDatabases != null },
+                { "comment", () => _comment, _comment != null }
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ListDatabasesOperation.cs:150]: let's not add delegate if it's not needed

File: src/MongoDB.Driver.Core/Core/Operations/ListIndexesOperation.cs [hunk 5]
```diff
@@ -72,6 +73,15 @@ public CollectionNamespace CollectionNamespace
             get { return _collectionNamespace; }
         }
 
+        /// <summary>
+        /// Gets or sets the comment.
+        /// </summary>
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ListIndexesOperation.cs:78]: `value` is missed

File: src/MongoDB.Driver.Core/Core/Operations/ListIndexesUsingCommandOperation.cs [hunk 6]
```diff
@@ -77,6 +76,15 @@ public CollectionNamespace CollectionNamespace
             get { return _collectionNamespace; }
         }
 
+        /// <summary>
+        /// Gets or sets the comment.
+        /// </summary>
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ListIndexesUsingCommandOperation.cs:81]: `value` is missed

File: src/MongoDB.Driver.Core/Core/Operations/RetryableWriteCommandOperationBase.cs [hunk 7]
```diff
@@ -58,6 +58,15 @@ public abstract class RetryableWriteCommandOperationBase : IWriteOperation<BsonD
         }
 
         // public properties
+        /// <summary>
+        /// Gets or sets the comment.
+        /// </summary>
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/RetryableWriteCommandOperationBase.cs:63]: `value` is missed

File: src/MongoDB.Driver/CreateManyIndexesOptions.cs [hunk 8]
```diff
@@ -24,10 +25,20 @@ namespace MongoDB.Driver
...
     {
         // private fields
+        private BsonValue _comment;
         private CreateIndexCommitQuorum _commitQuorum;
         private TimeSpan? _maxTime;
 
         // public properties
+        /// <summary>
+        /// Gets or sets the comment.
+        /// </summary>
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/CreateManyIndexesOptions.cs:35]: `value` is missed

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs:37]: I see you've created follow-up ticket: https://jira.mongodb.org/browse/CSHARP-4067
However the list of operation looks not 100% the same with what I provided above
NOTE: I would say we don't need to add $comment for legacy operations like `GeoSearch` (or operation that is already deprecated). 
Also not sure what operation you mean by `Create`

**Review by DmitryLukyanov — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs:37]: `DropCollection, DropDatabase` - I didn't find this in the follow up ticket, did I miss it?

[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs:37]: You are right, missed that as well, added.

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

**Pull request merged**